### ISSUE TITLE
Awk throws an error on OSX

### ISSUE THREAD

**Issue by m-czernek** [akinomyoga/cxxmatrix#3]
When running make on OSX, I get an error:

```
mbp:cxxmatrix osx$ awk -f glyph.awk glyph.def > glyph.inl
awk: calling undefined function or
 input record number 11, file glyph.def
 source line number 28
```

**Comment by akinomyoga** [comment]
Thank you very much for the report! I pushed a fix 761eeb0. Does the commit solve the issue?

**Comment by mimischi** [comment]
Running `make` the second time manages to compile the software.

```bash
$ git rev-parse HEAD
761eeb097205af1a21d0a78e6ee714454bf747dd
$ make
awk -f glyph.awk glyph.def > glyph.inl
awk: calling undefined function strtonum
 input record number 98, file glyph.def
 source line number 15
make: *** [glyph.inl] Error 2
$ make
c++ -std=c++17 -Wall -Wextra -Os -MD -MP -MF cxxmatrix.dep -c -o cxxmatrix.o cxxmatrix.cpp
cxxmatrix.cpp:27:15: warning: unused variable 'xmatrix_cell_power_max' [-Wunused-const-variable]
constexpr int xmatrix_cell_power_max = 10;
              ^
1 warning generated.
c++ -std=c++17 -Wall -Wextra -Os -o cxxmatrix cxxmatrix.o
```

Unfortunately the MacBook Air I'm using is struggling to keep up with the load :) Thanks for the software though!

**Comment by akinomyoga** [comment]
@mimischi  Thank you very much for the information!

> Unfortunately the MacBook Air I'm using is struggling to keep up with the load while running :)

The performance largely depends on the terminal emulators. For example, `iTerm2` is extremely slow according to the benchmark test by `alacritty`. Terminal.app and xterm are also slow. Maybe you can try it with faster terminal emulators such as urxvt (rxvt-unicode) or Alacritty (though I haven't tested with Alacritty).

**Comment by akinomyoga** [comment]
@mimischi Based on the error/warning messages in your report, I added additional fixes (eb0a326 and 82d7a71). Could you please try the latest version?

**Comment by m-czernek** [comment]
I am afraid I'm still unable to compile this:

```
mbp:cxxmatrix osx$ git rev-parse HEAD
385878cc4fa9b2529be779d9c412439488a5a50e
mbp:cxxmatrix osx$ make
awk -f glyph.awk glyph.def > glyph.inl
c++ -std=c++17 -Wall -Wextra -Os -MD -MP -MF cxxmatrix.dep -c -o cxxmatrix.o cxxmatrix.cpp
In file included from cxxmatrix.cpp:595:
./glyph.inl:78:2: error: illegal character encoding in character literal
{U'<EF>', 21, {127, 65, 65, 65, 65, 65, 127, }},
 ^
./glyph.inl:79:2: error: illegal character encoding in character literal
{U'<EF>', 21, {573695, 2072768, 104616, 104344, 37000, 1609860, 430211, }},
 ^
./glyph.inl:80:2: error: illegal character encoding in character literal
{U'<EF>', 21, {164354, 260859, 2015779, 1221155, 692770, 429602, 1238014, }},
 ^
./glyph.inl:81:2: error: illegal character encoding in character literal
{U'<EF>', 21, {526408, 1044472, 796815, 1841292, 659452, 592138, 583945, }},
 ^
./glyph.inl:82:2: error: illegal character encoding in character literal
{U'<EF>', 21, {33, 8161, 4159, 4136, 4136, 4134, 8161, }},
 ^
./glyph.inl:83:2: error: illegal character encoding in character literal
{U'<EF>', 21, {0, 0, 0, 0, 0, 0, 0, }},
 ^
./glyph.inl:84:2: error: illegal character encoding in character literal
{U'<EF>', 21, {0, 0, 0, 0, 0, 0, 0, }},
 ^
./glyph.inl:85:2: error: illegal character encoding in character literal
{U'<EF>', 21, {0, 0, 0, 0, 0, 0, 0, }},
 ^
./glyph.inl:86:2: error: illegal character encoding in character literal
{U'<EF>', 21, {0, 0, 0, 0, 0, 0, 0, }},
 ^
./glyph.inl:87:2: error: illegal character encoding in character literal
{U'<EF>', 21, {0, 0, 0, 0, 0, 0, 0, }},
 ^
./glyph.inl:88:2: error: illegal character encoding in character literal
{U'<EF>', 21, {0, 0, 0, 0, 0, 0, 0, }},
 ^
./glyph.inl:89:2: error: illegal character encoding in character literal
{U'<EF>', 21, {2088610, 49314, 53247, 49314, 1089696, 823440, 233102, }},
 ^
./glyph.inl:90:2: error: illegal character encoding in character literal
{U'<EF>', 21, {1771019, 722443, 1772543, 459530, 164490, 147979, 1719027, }},
 ^
./glyph.inl:91:2: error: illegal character encoding in character literal
{U'<EF>', 21, {825871, 862700, 1911052, 790525, 657674, 592137, 583912, }},
 ^
./glyph.inl:92:2: error: illegal character encoding in character literal
{U'<EF>', 21, {175, 160, 959, 7336, 168, 166, 161, }},
 ^
./glyph.inl:93:2: error: illegal character encoding in character literal
{U'<EF>', 21, {0, 0, 0, 0, 0, 0, 0, }},
 ^
./glyph.inl:94:2: error: illegal character encoding in character literal
{U'<EF>', 21, {0, 0, 0, 0, 0, 0, 0, }},
 ^
./glyph.inl:95:2: error: illegal character encoding in character literal
{U'<EF>', 21, {0, 0, 0, 0, 0, 0, 0, }},
 ^
./glyph.inl:96:2: error: illegal character encoding in character literal
{U'<EF>', 21, {0, 0, 0, 0, 0, 0, 0, }},
 ^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
20 errors generated.
make: *** [cxxmatrix.o] Error 1
```

I am unable to make your previous version too I'm afraid.

```
mbp:cxxmatrix osx$ git checkout 761eeb097205af1a21d0a78e6ee714454bf747dd
Note: checking out '761eeb097205af1a21d0a78e6ee714454bf747dd'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by performing another checkout.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -b with the checkout command again. Example:

  git checkout -b <new-branch-name>

HEAD is now at 761eeb0 glyph: avoid using GNU awk specific "or" and "lshift"
mbp:cxxmatrix osx$ git status
HEAD detached at 761eeb0
nothing to commit, working tree clean
mbp:cxxmatrix osx$ make
awk -f glyph.awk glyph.def > glyph.inl
awk: calling undefined function strtonum
 input record number 98, file glyph.def
 source line number 15
make: *** [glyph.inl] Error 2
```

**Comment by akinomyoga** [comment]
What is your locale? Can you provide me the results the following command?

```console
$ locale
```

Also can you try the following command instead of `make`?

```console
$ LC_ALL=C.UTF-8 make clean
$ LC_ALL=C.UTF-8 make all
```

**Comment by duraki** [comment]
@akinomyoga not OP but:

```
locale
LANG=""
LC_COLLATE="C"
LC_CTYPE="UTF-8"
LC_MESSAGES="C"
LC_MONETARY="C"
LC_NUMERIC="C"
LC_TIME="C"
LC_ALL=
```
  
The command you gave output the same error:

```
$ LC_ALL="C.UTF-8" make all
awk -f glyph.awk glyph.def > glyph.inl
c++ -std=c++17 -Wall -Wextra -Os -MD -MP -MF cxxmatrix.dep -c -o cxxmatrix.o cxxmatrix.cpp
In file included from cxxmatrix.cpp:595:
./glyph.inl:78:2: error: illegal character encoding in character literal
{U'<EF>', 21, {127, 65, 65, 65, 65, 65, 127, }},
 ^
./glyph.inl:79:2: error: illegal character encoding in character literal
{U'<EF>', 21, {573695, 2072768, 104616, 104344, 37000, 1609860, 430211, }},
 ^
./glyph.inl:80:2: error: illegal character encoding in character literal
{U'<EF>', 21, {164354, 260859, 2015779, 1221155, 692770, 429602, 1238014, }},
 ^
./glyph.inl:81:2: error: illegal character encoding in character literal
{U'<EF>', 21, {526408, 1044472, 796815, 1841292, 659452, 592138, 583945, }},
 ^
```

**Comment by akinomyoga** [comment]
@duraki Thank you very much! Do you also use macOS?

- It seems awk of @m-czernek and @duraki does not handle UTF-8 characters. Can you also check the results of the available locales with the following commands?

```console
$ locale -a
```

- Maybe awk in macOS does not support UTF-8. If you have gawk (GNU awk) in your system, could you try to rewrite `Makefile` to use `gawk` and see if it compiles with UTF-8 locales?

**Comment by akinomyoga** [comment]
I'm not sure if it solves the problem, but I added codes to detect UTF-8 locales and also GNU awk in Makefile in the commit 4fbee42.

**Comment by kirsis** [comment]
I'm also on macOS, and also had the illegal character encoding issue:

    ./glyph.inl:79:2: error: illegal character encoding in character literal
    ...

However, if I produce glyph.inl manually as follows, I can successfully compile:

    awk -f glyph.awk glyph.def | iconv -f ISO-8859-1 -t UTF-8 > glyph.inl

**Comment by akinomyoga** [comment]
@kirsis Thank you.

> awk -f glyph.awk glyph.def | iconv -f ISO-8859-1 -t UTF-8 > glyph.inl

It may compile, but it will not work as expected because the fragment `U'<EF>'` in the compile error message is not the character from ISO-8859-1 but a first byte of UTF-8 encoded Hankaku kanas. I find some users in HN reporting that it works without problems in macOS. Current my guess is that the problem depends on the set of locales installed in the system.

**Comment by saagarjha** [comment]
FWIW, installing a GNU awk (which https://github.com/akinomyoga/cxxmatrix/commit/4fbee4279484050070f40b30377f9ab4c67772eb detected) fixed the issue for me on macOS.

**Comment by akinomyoga** [comment]
@saagarjha Thank you! Maybe I need to add GNU awk to the requirements.

**Comment by duraki** [comment]
Pulling latest patch and then:

```
$ brew install gawk
$ brew unlink awk
$ brew link --overwrite gawk
$ make clean
$ make all
```

Worked fine for me. Thanks all!

**Comment by swills** [comment]
> @saagarjha Thank you! Maybe I need to add GNU awk to the requirements.

I had this same issue on FreeBSD until I installed gawk, so I'd say you do.

Also, you may want to rename your Makefile to GNUMakefile, see:

https://www.gnu.org/software/make/manual/make.html#Makefile-Names

**Comment by KhoomeiK** [comment]
I was able to fix it by find all 'ï' and replace with '\U000000EF' on MacOS. It seems this is the way all the other characters are encoded as well, could all 'ï' just be replaced with '\U000000EF' in the source?

**Comment by akinomyoga** [comment]
@duraki @swills I have already added GNU awk in requirements.

@swills Thank you. I renamed Makefile -> GNUmakefile.

@KhoomeiK `glyph.inl` is automatically generated by running AWK script `glyph.awk` for `glyph.def`. The file `glyph.inl` is not in the repository (I don't want to add machine generated codes to the git repository). Also, `U'\U000000EF'` is not what I intended. `ï` (0xEF) is just mis-encoded results by awk that does not understand UTF-8. It should be other characters which can be found in `glyph.def`.
